appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java

```diff
@@ -837,56 +838,58 @@ public synchronized void addJar(String jar, JarFile jarFile, File file)
             logger.log(Level.FINER, "addJar({0})", jar);
         }
 
-        // See IT 11417
-        super.addURL(getURL(file));
+        synchronized (jarFilesLock) {
+            // See IT 11417
+            super.addURL(getURL(file));
 
-        if ((jarPath != null) && (jar.startsWith(jarPath))) {
+            if ((jarPath != null) && (jar.startsWith(jarPath))) {
 
-            String jarName = jar.substring(jarPath.length());
-            while (jarName.startsWith("/")) {
-                jarName = jarName.substring(1);
+                String jarName = jar.substring(jarPath.length());
+                while (jarName.startsWith("/")) {
+                    jarName = jarName.substring(1);
+                }
+                jarNames.add(jarName);
             }
-            jarNames.add(jarName);
-        }
 
-        try {
-            // Register the JAR for tracking
+            try {
+                // Register the JAR for tracking
 
-            long lastModified = ((ResourceAttributes) resources.getAttributes(jar))
-                    .getLastModified();
+                long lastModified = ((ResourceAttributes) resources.getAttributes(jar))
+                        .getLastModified();
 
-            String[] result = new String[paths.length + 1];
-            for (int i = 0; i < paths.length; i++) {
-                result[i] = paths[i];
-            }
-            result[paths.length] = jar;
-            paths = result;
+                String[] result = new String[paths.length + 1];
+                for (int i = 0; i < paths.length; i++) {
+                    result[i] = paths[i];
+                }
+                result[paths.length] = jar;
+                paths = result;
```

Comment:
These 4 idioms for appending arrays are quite disturbing here. 

The design might stem from pre-1.4 era, where there was nothing similar to `CopyOnWriteArrayList<PathEntry>`, and is instead simulated with multiple arrays. But these should at least be refactored into form `paths = appendArray(paths, entry)`. And internally use `Arrays.copyOf(arr, arr.length+1)`.

Reply:
Correct, but the goal of this PR is not to update some already existing code constructs.